safehttp/mux.go

```diff
@@ -71,20 +71,42 @@ type ServeMux struct {
 	mux *http.ServeMux
 }
 
-func registerHandlers(mux *http.ServeMux, handlers map[string]map[string]handlerConfig) {
-	for pattern, handlersPerMethod := range handlers {
-		handlersPerMethod := handlersPerMethod
-		mux.HandleFunc(pattern, func(w http.ResponseWriter, r *http.Request) {
-			cfg, ok := handlersPerMethod[r.Method]
-			if !ok {
-				http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
-				return
-			}
-			processRequest(cfg, w, r)
-		})
+func registerHandlers(mux *http.ServeMux, handlers map[string]map[string]*handlerConfig) {
+	for pattern, methodToHandlerConfigMap := range handlers {
+		httpHandler := generateHttpHandlerFunc(methodToHandlerConfigMap)
+		mux.HandleFunc(pattern, httpHandler)
+	}
+}
+
+// generateHttpHandlerFunc converts a handlerConfig to a classic http.HandlerFunc.
+// We need to have this map because http.Handle handles requests for any HTTP request method per default,
+// which we want to avoid in safehttp.
+func generateHttpHandlerFunc(methodToHandlerConfig map[string]*handlerConfig) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		cfg, ok := methodToHandlerConfig[r.Method]
+		if !ok {
+			http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
+			return
+		}
+		processRequest(cfg, w, r)
 	}
 }
 
+// HttpHandlerForTransition generates a classic http.Handler using its ServeMuxConfig
+// (i.e. already installed Interceptors etc).
+// This is intended to be used during transition from using the http package to safehttp.
+func (s *ServeMuxConfig) HttpHandlerForTransition(method string, handler Handler, cfgs ...InterceptorConfig) http.Handler {
```

Comment:
As discussed: it'd make more sense to have a separate `HandlerFactory` or similar object which the `ServeMuxConfig` uses under the hood (or can be constructed from).

It feels weird to have to create a `ServeMuxConfig` and not use it to create a `Mux`.

Reply:
Agreed, will work on this once the other pieces in this PR look okay.

Reply:
I'd honestly have the `ServeMuxConfig` being used to instantiate the temporary factory for transition.

You will **eventually** use it to construct the Mux, but not during the transition phase. I think it is the right way to create a user journey towards the full framework adoption.

Reply:
Yeah, I also think that whatever we use here should be the thing that users use for both muxes and handlers (transition), which makes it a pretty important decision